### PR TITLE
Supports a JSON-format document service date

### PR DESCRIPTION
**Pull request by ndreynolds**
We've recently had some issues with parsing the service date because it's sometimes present as JSON and not in the serialized PHP format.  Normally it looks like this:

```
a:3:{s:15:\"serviceDateType\";s:1:\"1\";s:15:\"serviceDateFrom\";s:10:\"2015-08-01\";s:15:\"serviceDateThru\";s:10:\"2015-08-31\";}
```

But sometimes we're getting this instead (actual example):

```
{\"serviceDateType\":\"1\",\"serviceDateFrom\":\"2015-08-01\",\"serviceDateThru\":\"2015-08-31\"}
```

Instead of raising a parse error, this would first attempt to parse the input as JSON, then fall back to using `PHP.unserialize` if the input is not valid JSON.

### CODE REVIEW

File: lib/easybill/document.rb [hunk 1]
```diff
@@ -1,20 +1,19 @@
+require 'json'
 require 'php_serialize'
+require 'pry'
```
Comment (dziemba) [inline lib/easybill/document.rb]: :scream:

File: lib/easybill/document.rb [hunk 2]
```diff
@@ -34,5 +33,12 @@ def cancellation?
     def parse_date(date)
       Date.parse(date).strftime('%d.%m.%Y')
     end
+
+    def parse_service_date
+      return nil if service_date.nil?
+      JSON.parse(service_date)
```
Comment (dziemba) [inline lib/easybill/document.rb]: What about something like
```ruby
case service_date[0]
when 'a' then PHP...
when '{' then JSON...
else fail 'They introduced a 3rd format. Nice!'
end
```
instead? That's cleaner and a lot faster.

File: lib/easybill/document.rb (hunk 1 above)
Comment (ndreynolds) [inline lib/easybill/document.rb]: Oops, missed that. Fixed.

File: spec/easybill/document_spec.rb [hunk 3]
```diff
@@ -2,46 +2,75 @@
 
 describe Easybill::Document do
   describe '#service_period' do
-    subject { document.service_period }
+    subject { described_class.new('service_date' => service_date).service_period }
 
     context 'when a valid serviceDateString was returned' do
-      let(:document) { described_class.new('service_date' => "a:1:{s:17:\"serviceDateString\";s:23:\"01.05.2014 - 15.05.2014\";}") }
+      context 'PHP-serialized object' do
+        let(:service_date) { "a:1:{s:17:\"serviceDateString\";s:23:\"01.05.2014 - 15.05.2014\";}" }
```
Comment (dziemba) [inline spec/easybill/document_spec.rb]: Any reason for double-quoted string with escaped double-quotes?

File: lib/easybill/document.rb (hunk 2 above)
Comment (ndreynolds) [inline lib/easybill/document.rb]: I think the JSON parser is going to bail as soon as it sees the `a` anyway, though.  Even if it's a tiny bit faster, I feel like it's cleaner to delegate all the parsing to the respective libraries.  We would also have to handle at least a few additional cases (e.g., `N;` for null) that would make this more complicated.

Comment (dziemba) [inline lib/easybill/document.rb]: But... but... but my mum always told me not to use exceptions for control flow! :grinning: 

If you really prefer it this way, we can keep it. Do we have any feedback from easybill about this issue yet?

File: spec/easybill/document_spec.rb (hunk 3 above)
Comment (ndreynolds) [inline spec/easybill/document_spec.rb]: No, I removed them.  Was just how the examples were copied in.

File: lib/easybill/document.rb (hunk 2 above)
Comment (ndreynolds) [inline lib/easybill/document.rb]: Yeah, your mum has a good point :smile:.  I would've used something like `JSON.valid?(...)`, but it doesn't seem to exist.

Their documentation just says `serviceDate` is a string, so I guess it still technically works as described.  We could open a support ticket, though.  I also noticed they now have a REST API, though, that might be worth migrating over to at some point: https://www.easybill.de/api/.

Comment (dziemba) [inline lib/easybill/document.rb]: The new API looks promising! Then let's keep this PR as it is and make a ticket to switch to the new API.

**Comment by dziemba** [comment]
:mushroom:

**Pull request merged**